File: src/main/java/net/greenjab/fixedminecraft/registry/item/map_book/MapBookStateManager.java
package net.greenjab.fixedminecraft.registry.item.map_book;

import net.greenjab.fixedminecraft.FixedMinecraft;
import net.minecraft.resources.Identifier;
import net.minecraft.server.MinecraftServer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class MapBookStateManager {
    public ArrayList<Integer> currentBooks  = new ArrayList<>();
    public static MapBookStateManager INSTANCE = new MapBookStateManager();
    private final Map<Identifier, MapBookState> clientMapBooks = new HashMap<>();

       public MapBookState getMapBookState(MinecraftServer server, int id) {
        return server.getDataStorage().get(
                MapBookState.createStateType(getMapBookName(id)));
    }

    public void putMapBookState(MinecraftServer server, int id, MapBookState state) {
        server.getDataStorage().set(
                MapBookState.createStateType(getMapBookName(id)), state);
    }

    public MapBookState getClientMapBookState(int id) {
        return clientMapBooks.get(getMapBookName(id));
    }

    public void putClientMapBookState(int id, MapBookState state) {
        clientMapBooks.put(getMapBookName(id), state);
    }

    private Identifier getMapBookName(int mapId) {
        return FixedMinecraft.id("map_book/" + mapId);
    }

}

File: src/main/kotlin/net/greenjab/fixedminecraft/registry/item/map_book/MapBookStateManager.kt
package net.greenjab.fixedminecraft.registry.item.map_book

import net.greenjab.fixedminecraft.network.MapBookPlayer
import net.minecraft.datafixer.DataFixTypes
import net.minecraft.nbt.NbtCompound
import net.minecraft.registry.RegistryWrapper.WrapperLookup
import net.minecraft.server.MinecraftServer
import net.minecraft.world.PersistentState


class MapBookStateManager private constructor() {

    var currentBooks: ArrayList<Int> = ArrayList()

    val persistentStateType: PersistentState.Type<MapBookState>
        get() = PersistentState.Type(
            { throw IllegalStateException("Should never create an empty map saved data - but for map books") },
            { nbt: NbtCompound, lookup: WrapperLookup? ->
                INSTANCE.createMapBookState(
                    nbt
                )
            },
            DataFixTypes.SAVED_DATA_MAP_DATA
        )

    fun createMapBookState(nbt: NbtCompound): MapBookState {
        return MapBookState().fromNbt(nbt)
    }

    fun getMapBookState(server: MinecraftServer, id: Int): MapBookState? {
        return server.overworld.persistentStateManager.get(this.persistentStateType, this.getMapBookName(id))
    }

    fun putMapBookState(server: MinecraftServer, id: Int, state: MapBookState) {
        server.overworld.persistentStateManager[getMapBookName(id)] = state
    }

    fun getClientMapBookState(id: Int): MapBookState? {
        return clientMapBooks[getMapBookName(id)]
    }

    fun putClientMapBookState(id: Int, state: MapBookState) {
        clientMapBooks[getMapBookName(id)] = state
    }

    fun getMapBookName(mapId: Int): String {
        return "fixedminecraft_map_book_$mapId"
    }

    companion object {
        val INSTANCE: MapBookStateManager = MapBookStateManager()
        private val clientMapBooks: MutableMap<String, MapBookState> = HashMap()
    }
}
